debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -0,0 +1,159 @@
+ */
+public class ChangeEventQueue<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChangeEventQueue.class);
+
+    private final Duration pollInterval;
+    private final int maxBatchSize;
+    private final BlockingQueue<T> queue;
+    private final Metronome metronome;
+    private final Supplier<PreviousContext> loggingContextSupplier;
+
+    private volatile Throwable producerFailure;
+
+    private ChangeEventQueue(Duration pollInterval, int maxQueueSize, int maxBatchSize, Supplier<LoggingContext.PreviousContext> loggingContextSupplier) {
+        this.pollInterval = pollInterval;
+        this.maxBatchSize = maxBatchSize;
+        this.queue = new LinkedBlockingDeque<>(maxQueueSize);
+        this.metronome = Metronome.sleeper(pollInterval, Clock.SYSTEM);
+        this.loggingContextSupplier = loggingContextSupplier;
+    }
+
+    public static class Builder<T> {
+
+        private Duration pollInterval;
+        private int maxQueueSize;
+        private int maxBatchSize;
+        private Supplier<LoggingContext.PreviousContext> loggingContextSupplier;
+
+        public Builder<T> pollInterval(Duration pollInterval) {
+            this.pollInterval = pollInterval;
+            return this;
+        }
+
+        public Builder<T> maxQueueSize(int maxQueueSize) {
+            this.maxQueueSize = maxQueueSize;
+            return this;
+        }
+
+        public Builder<T> maxBatchSize(int maxBatchSize) {
+            this.maxBatchSize = maxBatchSize;
+            return this;
+        }
+
+        public Builder<T> loggingContextSupplier(Supplier<LoggingContext.PreviousContext> loggingContextSupplier) {
+            this.loggingContextSupplier = loggingContextSupplier;
+            return this;
+        }
+
+        public ChangeEventQueue<T> build() {
+            return new ChangeEventQueue<T>(pollInterval, maxQueueSize, maxBatchSize, loggingContextSupplier);
+        }
+    }
+
+    public void enqueue(T record) {
+        LoggingContext.PreviousContext previousContext = loggingContextSupplier.get();
+
+        try {
+            queue.put(record);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Placed source record '{}' into queue", record);
+            }
+        } catch (InterruptedException e) {
+            LOGGER.debug("received interrupt request");
+            // clear the interrupted status
+            Thread.interrupted();
+        } finally {
+            previousContext.restore();
+        }
+    }
+
+    public List<T> poll() {
+        LoggingContext.PreviousContext previousContext = loggingContextSupplier.get();
+
+        try {
+            LOGGER.debug("polling records...");
+            List<T> records = new ArrayList<>();
+            final Timer timeout = Threads.timer(Clock.SYSTEM, Temporals.max(pollInterval, ConfigurationDefaults.RETURN_CONTROL_INTERVAL));
+            while (!timeout.expired() && queue.drainTo(records, maxBatchSize) == 0) {
```

Comment:
Could we add also Thread.isInterrupted() into the condition? As the interrupt might arrive even when the thread is not blocked.

Reply:
I don't think it adds much benefit. The control flow will very quickly arrive at the `metronome.pause()` call, which is interruptable. I.e. checking the interrupted flag in the condition would abort the loop just a tiiiny bit earlier.